src/log_unstable.rs

```diff
@@ -150,20 +157,26 @@ impl Unstable {
         let after = ents[0].index;
         if after == self.offset + self.entries.len() as u64 {
             // after is the next index in the self.entries, append directly
-            self.entries.extend_from_slice(ents);
         } else if after <= self.offset {
             // The log is being truncated to before our current offset
             // portion, so set the offset and replace the entries
             self.offset = after;
             self.entries.clear();
-            self.entries.extend_from_slice(ents);
+            self.entries_size = 0;
         } else {
             // truncate to after and copy to self.entries then append
             let off = self.offset;
             self.must_check_outofbounds(off, after);
+            for e in &self.entries[(after - off) as usize..] {
+                self.entries_size -= e.compute_size() as usize;
+            }
             self.entries.truncate((after - off) as usize);
-            self.entries.extend_from_slice(ents);
         }
+        self.entries.extend_from_slice(ents);
+        self.entries_size += ents
+            .iter()
+            .map(|ent| ent.compute_size() as usize)
```

Comment:
How about counting the data size and context size directly?

Reply:
Done